dali/pipeline/operator/builtin/external_source.cc

```diff
@@ -20,38 +20,71 @@ namespace dali {
 template<>
 void ExternalSource<CPUBackend>::RunImpl(HostWorkspace &ws) {
   std::list<uptr_tl_type> tensor_list_elm;
+  std::list<uptr_tv_type> tensor_vector_elm;
   std::list<uptr_cuda_event_type> internal_copy_to_storage;
+  ZeroCopyInfo copy_info;
   {
     std::unique_lock<std::mutex> busy_lock(busy_m_);
-    tensor_list_elm = tl_data_.PopFront();
-    internal_copy_to_storage = copy_to_storage_events_.PopFront();
+    copy_info = zero_copy_.front();
+    zero_copy_.pop_front();
+    if (copy_info.is_tensor_vector && copy_info.is_zero_copy) {
+      tensor_vector_elm = tv_data_.PopFront();
+    } else {
+      tensor_list_elm = tl_data_.PopFront();
+    }
+    if (!copy_info.is_zero_copy) {
+      internal_copy_to_storage = copy_to_storage_events_.PopFront();
```

Comment:
Either it's somewhere hidden or you don't use this at all.

Reply:
We always adding it inside `SetDataSourceHelper` to `copy_to_storage_events_` but maybe I can just limit that to the GPU variant.

Reply:
Done